src/db/schemas/user.ts

```diff
@@ -1,36 +1,84 @@
+import * as bcrypt from 'bcrypt-nodejs';
 import * as mongoose from 'mongoose';
 
-const UserSchema = {
-  confirmed: {
-    type: Boolean,
-  },
-  email: {
-    required: true,
-    type: String,
-  },
-  image: {
-    type: String,
-  },
-  interfaceLang: {
-    default: 'ru',
-    required: true,
-    type: String,
-  },
-  name: {
-    required: true,
-    type: String,
-  },
-  password: {
-    required: true,
-    type: String,
-  },
-  subscription: {
-    default: true,
-    required: true,
-    type: Boolean,
+import { IUser } from '../../interfaces/index';
+import { sendMail } from '../../utils';
+
+export interface IUserModel extends IUser, mongoose.Document {
+  comparePassword(candidatePassword: string, callback: any): any;
+}
+
+const UserSchema = new mongoose.Schema(
+  {
+    confirmed: {
+      default: false,
+      type: Boolean
+    },
+    email: {
+      required: true,
+      type: String,
+      unique: true
+    },
+    image: {
+      default: '',
+      type: String
+    },
+    interfaceLang: {
+      default: 'ru',
+      required: true,
+      type: String
+    },
+    name: {
+      required: true,
+      type: String
+    },
+    password: {
+      required: true,
+      type: String
+    },
+    subscription: {
+      default: true,
+      required: true,
+      type: Boolean
+    }
   },
-};
+  { timestamps: true }
+);
+
+UserSchema.pre('save', function(next) {
+  const user = this;
+
+  bcrypt.genSalt(10, (err, salt) => {
+    if (err) {
+      return next(err);
+    }
+
+    bcrypt.hash(user.password, salt, null, (error: any, hash: any) => {
+      if (error) {
+        return next(error);
+      }
 
-const User = mongoose.model('User', new mongoose.Schema(UserSchema));
+      const tomorrow = new Date();
+      tomorrow.setDate(tomorrow.getDate() + 1);
+
+      user.password = hash;
+      next();
+    });
+  });
+});
+
+UserSchema.post('save', function() {
+  const user = this;
+  sendMail(this.email, this.name);
```

Comment:
Is this action sync? Doesn't it take a callback?

Reply:
As I understand it takes callback if we have sequence of "post" hooks.
I found it here http://mongoosejs.com/docs/middleware.html . As I think we can use post middleware both for async and sync actions.

Reply:
I mean how the post method knows when the sendMail execution is ended?